### PR TITLE
Only reprocess batch tallies when relevant contest fields change

### PR DESCRIPTION
**Pull request by jonahkagan**
Fixes https://github.com/votingworks/arlo/issues/1977

Previously, we were reprocessing all jurisdictions' batch tallies files whenever we made a request to update contests. This can be pretty disruptive because it affects every jurisdiction.

There are a few cases that we know don't actually impact the batch tallies at all, so we can skip reprocessing:
- When the audit admin doesn't change the contests at all (e.g. if they click "Save & Next" on the contests form without changing anything
- When the audit admin only changes the contest's jurisdiction list. Since batch tallies are specific to a single jurisdiction, adding or removing a jurisdiction will have no impact on other jurisdiction's batch tallies.

### CODE REVIEW

**Review by arsalansufi — APPROVED** [review]
LGTM! Wishing I'd reviewed this last night 😅 as it would've come in handy today per all the reprocessing messages in #arlo-activity

File: server/api/contests.py [hunk 1]
```diff
@@ -318,6 +344,24 @@ def create_or_update_all_contests(election: Election):
                 "Cannot update contests while batch tallies file is being processed."
             )
 
+        previous_contests = [
+            serialize_contest(contest) for contest in election.contests
+        ]
+
+    for contest in election.contests:
+        db_session.delete(contest)
```
Comment (kshen0) [inline server/api/contests.py:352]: Why change this from the previous `Contest.query.filter_by(election_id=election.id).delete()`?

**Review by kshen0 — APPROVED** [review]

File: server/api/contests.py (hunk 1 above)
Comment (jonahkagan) [reply server/api/contests.py:352]: Accessing `election.contests` on L348 just above causes the contests to be loaded into the ORM's cache, which caused a stale data error when later trying to delete and then commit the new contests. I honestly don't understand the error, but this fixed it.

File: server/api/contests.py [hunk 2]
```diff
@@ -287,28 +287,54 @@ def set_contest_metadata(election: Election):
...
+# affects every jurisdiction.
+def should_reprocess_batch_tallies(
+    previous_contests: List[JSONDict], new_contests: List[JSONDict]
+) -> bool:
+    if len(previous_contests) != len(new_contests):
+        return True
+
+    def normalize_contest(contest):
+        contest = contest.copy()
+        # This field isn't sent from the client, its computed on the backend, so
```
[the remark this answers is not in the document]
Comment (jonahkagan) [reply server/api/contests.py:304]: Whoops, didn't get the email notification about this comment before merging 🤦 Doesn't feel worth a PR to fix

[the remark this answers is not in the document]
Comment (kshen0) [reply server/api/contests.py:304]: No problem, I think I actually submitted 1 minute after your merge

**Pull request merged**